### PR TITLE
[EOS-15552] Implement API for endpoint address decode

### PR DESCRIPTION
**Pull request by VinothVijayakumar-dev**
- API to decode node and port number from given ep_name

Signed-off-by: Vinoth.V <vinoth.v@seagate.com>

### CODE REVIEW

File: net/libfab/libfab.c [hunk 1]
```diff
@@ -95,6 +123,8 @@ static int libfab_dom_init(struct m0_net_xprt *xprt, struct m0_net_domain *dom)
 	* fab_hints->caps = FI_MSG;
 	* fab_hints->fabric_attr->prov_name = "verbs";
 	*/
+	libfab_ep_addr_decode(NULL, node, port);
```
Comment (upendrapatwardhan) [inline net/libfab/libfab.c:126]: Passing the first parameter as NULL will return what values in node and port?

File: net/libfab/libfab.c [hunk 2]
```diff
@@ -70,12 +70,43 @@ enum m0_fab__mr_params {
...
+
+	M0_PRE(ep_name != NULL);
+
+	if (cp == NULL || n == 0 )
+		return M0_ERR(-EINVAL);
+
+	M0_ENTRY("ep_name=%s", ep_name);
+
+	memcpy(node, ep_name, n);
+	node[n] = 0;
```
Comment (jamohite) [inline net/libfab/libfab.c:90]: should it be '\0'  here are and at line no 95 ?

File: Makefile.am [hunk 3]
```diff
@@ -187,6 +187,10 @@ motr_libmotr_la_LIBADD    = @MATH_LIBS@ @PTHREAD_LIBS@ @AIO_LIBS@ @RT_LIBS@ \
                             @YAML_LIBS@ @PROFILER_LIBS@ @UUID_LIBS@ @GALOIS_LIBS@ \
                             @DL_LIBS@ @CASSANDRA_LIBS@ @UV_LIBS@
 
+if ENABLE_LIBFAB
```
Comment (huanghua78) [inline Makefile.am:190]: Please rebase to libfabric_integration.

I think this chanhe is already merged.

File: net/libfab/libfab.c [hunk 4]
```diff
@@ -0,0 +1,525 @@
...
+			FI_REMOTE_READ | FI_REMOTE_WRITE),
+	/** Fabric memory offset. */
+	FAB_MR_OFFSET  = 0,
+	/** Fabric memory flag. */
+	FAB_MR_FLAG    = 0,
+	/** Key used for memory registration. */
+	FAB_MR_KEY     = 0XABCD,
+};
+
+static int libfab_ep_addr_decode(const char *ep_name, char *node, char *port);
```
Comment (huanghua78) [inline net/libfab/libfab.c:73]: Remove this.

File: net/libfab/libfab.c [hunk 5]
```diff
@@ -0,0 +1,525 @@
...
+
+/* Used to take the ip and port from the given end point */
+static int libfab_ep_addr_decode(const char *ep_name, char *node, char *port)
+{
+	char     *cp = strchr(ep_name, ':');
```
Comment (huanghua78) [inline net/libfab/libfab.c:78]: Move this to the position after "M0_PRE(ep_name != NULL);"

File: net/libfab/libfab.c [hunk 6]
```diff
@@ -0,0 +1,525 @@
...
+	size_t    n  = cp - ep_name;
+	int       rc = 0;
+
+	M0_PRE(ep_name != NULL);
+
+	if (cp == NULL || n == 0 )
+		return M0_ERR(-EINVAL);
```
Comment (huanghua78) [inline net/libfab/libfab.c:85]: I would prefer something like this:
```
	char     *cp ;
	size_t    n;
	int       rc = 0;

        M0_PRE(ep_name != NULL);

	cp = strchr(ep_name, ':');
        if (cp == NULL)
                return M0_ERR(-EINVAL);
	n  = cp - ep_name;
	if (n == 0 )
		return M0_ERR(-EINVAL);
```

Don't operate on invalid variables.

File: net/libfab/libfab.c [hunk 7]
```diff
@@ -0,0 +1,525 @@
...
+
+	M0_ENTRY("ep_name=%s", ep_name);
+
+	memcpy(node, ep_name, n);
```
Comment (huanghua78) [inline net/libfab/libfab.c:89]: How do we make sure the 'node' buffer is large enough?

File: net/libfab/libfab.c [hunk 8]
```diff
@@ -0,0 +1,525 @@
...
+	node[n] = 0;
+
+	++cp;
+	n=strlen(cp);
+	memcpy(port, cp, n);
```
Comment (huanghua78) [inline net/libfab/libfab.c:94]: How do we make sure the 'port' buffer is large enough?

File: net/libfab/libfab.c [hunk 9]
```diff
@@ -0,0 +1,525 @@
...
+	FAB_MR_OFFSET  = 0,
+	/** Fabric memory flag. */
+	FAB_MR_FLAG    = 0,
+	/** Key used for memory registration. */
+	FAB_MR_KEY     = 0XABCD,
+};
+
+static int libfab_ep_addr_decode(const char *ep_name, char *node, char *port);
+
+/* Used to take the ip and port from the given end point */
```
Comment (huanghua78) [inline net/libfab/libfab.c:75]: Please use "Doxygen" style comments:
```
/**
 * This is a function to ...
 * Add comment about its parameters and return value.
 */
```

Comment (huanghua78) [reply net/libfab/libfab.c:75]: It's better to have some examples.
What is the format of the EP.

File: net/libfab/libfab.c (hunk 7 above)
Comment (upendrapatwardhan) [reply net/libfab/libfab.c:89]: I think we should also pass buffer length as an argument and pre validate if it is large enough

File: net/libfab/libfab.c [hunk 10]
```diff
@@ -0,0 +1,525 @@
...
+}
+
+
+/** Used as m0_net_xprt_ops::xo_dom_init(). */
+static int libfab_dom_init(struct m0_net_xprt *xprt, struct m0_net_domain *dom)
+{
+	struct m0_fab__dom_param *fab_dom;
+	struct fi_info           *fab_hints;
+	int 			  rc;
+	char                      node[16];
```
Comment (upendrapatwardhan) [inline net/libfab/libfab.c:107]: Buffer size should support for ipv6 addresses also

File: net/libfab/libfab.c [hunk 11]
```diff
@@ -92,12 +92,73 @@ M0_INTERNAL void m0_net_libfab_fini(void)
...
+/**
+ * Used to take the ip and port from the given end point
+ * ep_name : endpoint address from domain
+ * node    : copy ip address from ep_name
+ * port    : copy port number from ep_name
+ * Example of ep_name IPV4 192.168.0.1:4235
+ *                    IPV6 [4002:db1::1]:4235
+ */
+static int libfab_ep_addr_decode(const char *ep_name, char *node,
+			  size_t nodeSize, char *port, size_t portSize)
```
Comment (upendrapatwardhan) [inline net/libfab/libfab.c:104]: Alignment

File: net/libfab/libfab.c [hunk 12]
```diff
@@ -92,12 +92,73 @@ M0_INTERNAL void m0_net_libfab_fini(void)
...
+
+	if( ep_name[0] == '[' ) {
+		/* IPV6 pattern */
+		cp = strchr(ep_name, ']');
+		if (cp == NULL)
+			return M0_ERR(-EINVAL);
+
+		ep_name++;
+		n = cp - ep_name;
+                if (n == 0 )
```
Comment (huanghua78) [inline net/libfab/libfab.c:122]: Alignment.

File: net/libfab/libfab.c [hunk 13]
```diff
@@ -92,12 +92,73 @@ M0_INTERNAL void m0_net_libfab_fini(void)
...
+                        return M0_ERR(-EINVAL);
+
```
Comment (huanghua78) [inline net/libfab/libfab.c:124]: At this moment, we must check the ":",
```
cp++;
if (*cp != ':')
        return M0_ERR(-EINVAL);
cp++;
```

File: net/libfab/libfab.c [hunk 14]
```diff
@@ -92,12 +92,73 @@ M0_INTERNAL void m0_net_libfab_fini(void)
...
+	return M0_RC(rc);
+}
+
 /** Used as m0_net_xprt_ops::xo_dom_init(). */
 static int libfab_dom_init(struct m0_net_xprt *xprt, struct m0_net_domain *dom)
 {
 	struct m0_fab__dom_param *fab_dom;
 	struct fi_info           *fab_hints;
 	int 			  rc;
+	char                      node[40];
```
Comment (huanghua78) [inline net/libfab/libfab.c:159]: Maybe we need to use this one:
```
./netinet/in.h:234:#define INET6_ADDRSTRLEN 46
```

File: net/libfab/libfab.c [hunk 15]
```diff
@@ -92,12 +92,73 @@ M0_INTERNAL void m0_net_libfab_fini(void)
 	*/
 }
 
+/**
+ * Used to take the ip and port from the given end point
+ * ep_name : endpoint address from domain
+ * node    : copy ip address from ep_name
+ * port    : copy port number from ep_name
+ * Example of ep_name IPV4 192.168.0.1:4235
```
Comment (huanghua78) [inline net/libfab/libfab.c:100]: Please note, the 'sock' xprt recognize the lnet ep format.
Does libfab need the same?
I think we it's better to use the lnet compatible format. So, a lot of code can keep the same.

Please discuss this with @nikitadanilov

File: net/libfab/libfab.c [hunk 16]
```diff
@@ -92,12 +99,239 @@ M0_INTERNAL void m0_net_libfab_fini(void)
...
+	* that the port is not reserved), followed by 5 bits of (portal - 30),
+	* so that portal must be in the range 30..61.
+	*/
+	if (tmid >= 1024 || (portal - 30) >= 32)
+		return M0_ERR_INFO(-EPROTO,
+			"portal: %u, tmid: %u", portal, tmid);
+
+	portnum  = htons(tmid | (1 << 10) | ((portal - 30) << 11));
+	//sscanf(portnum,"%d",port);
+	sprintf(port,"%d",portnum);
```
Comment (huanghua78) [inline net/libfab/libfab.c:171]: Add spaces between params.
```
sprintf(port, "%d", portnum);
```

File: net/libfab/libfab.c [hunk 17]
```diff
@@ -92,12 +99,239 @@ M0_INTERNAL void m0_net_libfab_fini(void)
...
+        } else {
+                at++;
+	        if (at == NULL) 
+                        return M0_ERR(-EINVAL);
+	        M0_PRE(portSize >= (strlen(at)+1));
+                memcpy(port,at,(strlen(at)+1));
+        }
+	M0_PRE(nodeSize >= (at - ep_name));
+        memcpy(node, ep_name, ((at - ep_name)-1));
+        return 0;
```
Comment (huanghua78) [inline net/libfab/libfab.c:218]: Mixed and wrong indentation.

**Review by huanghua78 — APPROVED** [review]
Looks fine.

Minor indentation issue.

**Review by upendrapatwardhan — APPROVED** [review]

**Comment by huanghua78** [comment]
Is this PR still needed? @VinothVijayakumar-dev

**Pull request merged**